File: examples/canonical-analytics/python/src/analytics_workbench/filters.py
from dataclasses import dataclass, field
from datetime import date
from typing import Optional
from fastapi import Request

@dataclass
class FilterState:
    date_from: Optional[date] = None
    date_to: Optional[date] = None
    services: list[str] = field(default_factory=list)
    severity: list[str] = field(default_factory=list)
    environment: list[str] = field(default_factory=list)

def _parse_multi(request: Request, key: str) -> list[str]:
    # handles repeated query params like ?services=a&services=b
    # and comma-separated like ?services=a,b
    values = request.query_params.getlist(key)
    res = []
    for v in values:
        for p in v.split(","):
            p = p.strip()
            if p:
                res.append(p)
    return res

def parse_filter_state(request: Request) -> FilterState:
    df_str = request.query_params.get("date_from")
    dt_str = request.query_params.get("date_to")
    
    date_from = date.fromisoformat(df_str) if df_str else None
    date_to = date.fromisoformat(dt_str) if dt_str else None
    
    services = _parse_multi(request, "services")
    severity = _parse_multi(request, "severity")
    environment = _parse_multi(request, "environment")
    
    return FilterState(
        date_from=date_from,
        date_to=date_to,
        services=services,
        severity=severity,
        environment=environment
    )

File: examples/canonical-analytics/elixir/analytics_workbench/lib/analytics_workbench/filters.ex
defmodule AnalyticsWorkbench.Filters do
  defstruct date_from: nil, date_to: nil,
            services: [], severity: [], environment: []

  def parse(params) do
    %__MODULE__{
      date_from: Map.get(params, "date_from"),
      date_to: Map.get(params, "date_to"),
      services: parse_multi(params, "services"),
      severity: parse_multi(params, "severity"),
      environment: parse_multi(params, "environment"),
    }
  end

  defp parse_multi(params, key) do
    case Map.get(params, key) do
      nil -> []
      val when is_list(val) -> val
      val when is_binary(val) ->
        if String.contains?(val, ",") do
           String.split(val, ",") |> Enum.reject(&(&1 == ""))
        else
           [val] |> Enum.reject(&(&1 == ""))
        end
    end
  end
end
